network: enabling IPv6 when using host platform results in uncaught error

### The problem

I attempted to build a host platform esphome config and I left in this:

``` yaml
network:
  enable_ipv6: true
```

I didn't need it, it was just left over from a config I ported from an ESP32. I assume that this feature is unsupported, as I just get this error on compilation:

```
src/esphome/components/socket/socket.cpp: In function 'size_t esphome::socket::format_sockaddr_to(const sockaddr*, socklen_t, std::span<char, 46>)':
src/esphome/components/socket/socket.cpp:64:25: error: 'const struct in6_addr' has no member named 'un'
   64 |     if (addr->sin6_addr.un.u32_addr[0] == 0 && addr->sin6_addr.un.u32_addr[1] == 0 &&
      |                         ^~
src/esphome/components/socket/socket.cpp:64:64: error: 'const struct in6_addr' has no member named 'un'
   64 |     if (addr->sin6_addr.un.u32_addr[0] == 0 && addr->sin6_addr.un.u32_addr[1] == 0 &&
      |                                                                ^~
src/esphome/components/socket/socket.cpp:65:25: error: 'const struct in6_addr' has no member named 'un'
   65 |         addr->sin6_addr.un.u32_addr[2] == htonl(0xFFFF) &&
      |                         ^~
src/esphome/components/socket/socket.cpp:66:45: error: 'const struct in6_addr' has no member named 'un'
   66 |         esphome_inet_ntop4(&addr->sin6_addr.un.u32_addr[3], buf.data(), buf.size()) != nullptr) {
      |                                             ^~
Compiling .pioenvs/ctrl-one/src/esphome/core/gpio.o
*** [.pioenvs/ctrl-one/src/esphome/components/socket/socket.o] Error 1
```

I expect either this to work, or be caught at the YAML validation stage.

### Which version of ESPHome has the issue?

2026.3.0-dev

### What type of installation are you using?

pip

### What platform are you using?

Host

### Component causing the issue

network

### YAML Config

```yaml
esphome:
  name: host-demo

host:
  mac_address: 0F:70:E8:50:BB:DB

logger:
  level: DEBUG

api:
  encryption:
    key: !secret encryption_key

network:
  enable_ipv6: true

button:
  - platform: restart
    name: Restart

binary_sensor:
  - platform: status
    id: system_status

time:
  - platform: host
    id: sntp_time
    timezone: Europe/Amsterdam

sensor:
  - platform: uptime
    name: Uptime
    id: uptime_sensor

lvgl:
  buffer_size: 25%
  on_draw_end:
    - component.update: goodisplay

display:
  - platform: sdl
    id: goodisplay
    auto_clear_enabled: false
    update_interval: never
    dimensions:
      width: 300
      height: 400

output:
  - platform: gpio
    pin: GPIO0
    id: screen_led

touchscreen:
  - platform: sdl
    id: touchscreen_ft
    on_touch:
      - lambda: |-
          ESP_LOGI("cal", "x=%d, y=%d, x_raw=%d, y_raw=%0d",
              touch.x,
              touch.y,
              touch.x_raw,
              touch.y_raw
              );
```

### Anything in the logs that might be useful for us?

```txt

```

### Additional information

_No response_

--- a/esphome/components/socket/socket.cpp
+++ b/esphome/components/socket/socket.cpp
@@ -59,8 +59,14 @@ size_t format_sockaddr_to(const struct sockaddr *addr_ptr, socklen_t len, std::s
 #if USE_NETWORK_IPV6
   else if (addr_ptr->sa_family == AF_INET6 && len >= sizeof(sockaddr_in6)) {
     const auto *addr = reinterpret_cast<const struct sockaddr_in6 *>(addr_ptr);
-#ifndef USE_SOCKET_IMPL_LWIP_TCP
-    // Format IPv4-mapped IPv6 addresses as regular IPv4 (not supported on ESP8266 raw TCP)
+#ifdef USE_HOST
+    // Format IPv4-mapped IPv6 addresses as regular IPv4 (POSIX layout, no LWIP union)
+    if (IN6_IS_ADDR_V4MAPPED(&addr->sin6_addr) &&
+        esphome_inet_ntop4(&addr->sin6_addr.s6_addr[12], buf.data(), buf.size()) != nullptr) {
+      return strlen(buf.data());
+    }
+#elif !defined(USE_SOCKET_IMPL_LWIP_TCP)
+    // Format IPv4-mapped IPv6 addresses as regular IPv4 (LWIP layout)
     if (addr->sin6_addr.un.u32_addr[0] == 0 && addr->sin6_addr.un.u32_addr[1] == 0 &&
         addr->sin6_addr.un.u32_addr[2] == htonl(0xFFFF) &&
         esphome_inet_ntop4(&addr->sin6_addr.un.u32_addr[3], buf.data(), buf.size()) != nullptr) {

--- a/tests/components/socket/common.yaml
+++ b/tests/components/socket/common.yaml
@@ -1,0 +1,11 @@
+substitutions:
+  network_enable_ipv6: "false"
+
+socket:
+
+wifi:
+  ssid: MySSID
+  password: password1
+
+network:
+  enable_ipv6: ${network_enable_ipv6}

--- a/tests/components/socket/test-ipv6.esp32-idf.yaml
+++ b/tests/components/socket/test-ipv6.esp32-idf.yaml
@@ -1,0 +1,4 @@
+substitutions:
+  network_enable_ipv6: "true"
+
+<<: !include common.yaml

--- a/tests/components/socket/test-ipv6.host.yaml
+++ b/tests/components/socket/test-ipv6.host.yaml
@@ -1,0 +1,4 @@
+socket:
+
+network:
+  enable_ipv6: true

--- a/tests/components/socket/test.esp32-idf.yaml
+++ b/tests/components/socket/test.esp32-idf.yaml
@@ -1,0 +1,1 @@
+<<: !include common.yaml

--- a/tests/components/socket/test.host.yaml
+++ b/tests/components/socket/test.host.yaml
@@ -1,0 +1,3 @@
+socket:
+
+network:

